src/Blocks/Block.php

```diff
@@ -0,0 +1,253 @@
+<?php
+
+namespace Notion\Blocks;
+
+use DateTimeImmutable;
+use Notion\Common\Date;
+
+/**
+ * @psalm-type BlockJson = array{
+ *      type: string,
+ *      id: string,
+ *      created_time: string,
+ *      last_edited_time: string,
+ *      archived: bool,
+ *      has_children: bool,
+ * }
+ */
+class Block
```

Comment:
Regarding the hierarchy, I've tried to avoid some huge code duplication.

Every block has the basic properties: `id`, `createdTime`, `lastEditedTime`, `archived`, `hasChildren` and `type`. Those were abstracted by the class `Block`.

I wanted to avoid class inheritance, by preferring composition (not sure if I did it right).

Take the `Paragraph` block for example, but it applies to all other types of block:

```php
class Paragraph implements BlockInterface
{
    // common block details (id, createdTime, ...)
    private Block $block;

    // Paragraph specific properties
    private array $text;
    private array $children;
    
    // interface methods
    public function block(): Block { /* ... */ }
    public function toArray(): array { /* ... */ }
    public static function fromArray(array $input): self { /* ... */ }
}
```

With the explanation, is it still unclear? Would you suggest a different approach?

Reply:
> I wanted to avoid class inheritance, by preferring composition (not sure if I did it right).

That works, but then IMO `toArray()` should probably handle all of this, no? `block()` should be unnecessary, in theory.